### PR TITLE
feat: implement keyword and hybrid search for Weaviate provider

### LINKED ISSUE #3010: Implement `v1/vector_stores/{vector_store_id}/search` for Weaviate
### 🚀 Describe the new functionality needed

Implement `v1/vector_stores/{vector_store_id}/search` for Weaviate

### 💡 Why is this needed? What if we don't build it?

openai compat

### Other thoughts

_No response_

### PR DESCRIPTION
**Pull request by ChristianZaccaria**
# What does this PR do?

- This PR implements keyword and hybrid search for Weaviate DB based on its inbuilt functions.
- Added unit tests for the WeaviateIndex class.
- Added fixtures to conftest.py for Weaviate.
- Enabled integration tests for remote Weaviate on all 3 search modes.

Closes #3010 

## Test Plan

Unit tests and integration tests should pass on this PR.

### CODE REVIEW

File: tests/unit/providers/vector_io/conftest.py [hunk 1]
```diff
@@ -333,6 +335,78 @@ async def qdrant_vec_index(qdrant_vec_db_path, embedding_dimension):
     await index.delete()
 
 
+@pytest.fixture
+def weaviate_vec_db_path():
+    return "localhost:8080"
```
Comment (2000krysztof) [inline tests/unit/providers/vector_io/conftest.py]: This might be a bit overkill

File: tests/unit/providers/vector_io/conftest.py [hunk 2]
```diff
@@ -333,6 +335,78 @@ async def qdrant_vec_index(qdrant_vec_db_path, embedding_dimension):
...
+
+
+@pytest.fixture
+async def weaviate_vec_index(weaviate_vec_db_path, embedding_dimension):
```
Comment (2000krysztof) [inline tests/unit/providers/vector_io/conftest.py]: weaviate_vec_db_path is not being called in this function at all, it might be worth removing it from the paramaters for better readability.

**Review by 2000krysztof — CHANGES_REQUESTED** [review]

File: tests/unit/providers/vector_io/conftest.py (hunk 1 above)
Comment (ChristianZaccaria) [reply tests/unit/providers/vector_io/conftest.py]: For Weaviate I think this is fine, especially for testing. This is setup here: https://github.com/llamastack/llama-stack/blob/main/.github/workflows/integration-vector-io-tests.yml#L59 - I'll have another look at this, thanks!

Comment (2000krysztof) [reply tests/unit/providers/vector_io/conftest.py]: No I mean it is overkill to define a separate fixture if it is only used in two places.

Comment (ChristianZaccaria) [reply tests/unit/providers/vector_io/conftest.py]: True, mostly I left it to follow the pattern. But, you're right. I'll remove it, thanks!

File: llama_stack/providers/remote/vector_io/weaviate/weaviate.py [hunk 3]
```diff
@@ -47,7 +48,7 @@
 class WeaviateIndex(EmbeddingIndex):
     def __init__(
         self,
-        client: weaviate.Client,
+        client: weaviate.WeaviateClient,
```
Comment (franciscojavierarceo) [inline llama_stack/providers/remote/vector_io/weaviate/weaviate.py:51]: Why this change?

File: llama_stack/providers/remote/vector_io/weaviate/weaviate.py [hunk 4]
```diff
@@ -136,7 +153,46 @@ async def query_keyword(
...
+            score_threshold: Minimum similarity score threshold
+        Returns:
+            QueryChunksResponse with chunks and scores
+        """
+        log.info(f"WEAVIATE KEYWORD SEARCH CALLED: query='{query_string}', k={k}, threshold={score_threshold}")
+        sanitized_collection_name = sanitize_collection_name(self.collection_name, weaviate_format=True)
+        collection = self.client.collections.get(sanitized_collection_name)
+
+        # Perform BM25 keyword search on chunk_content field
+        results = collection.query.bm25(
```
Comment (franciscojavierarceo) [inline llama_stack/providers/remote/vector_io/weaviate/weaviate.py]: Need to add try and except block if weaviate client fails

File: llama_stack/providers/remote/vector_io/weaviate/weaviate.py [hunk 5]
```diff
@@ -147,7 +203,62 @@ async def query_hybrid(
...
+
+        # Ranked (RRF) reranker fusion type
+        if reranker_type == RERANKER_TYPE_RRF:
+            rerank = HybridFusion.RANKED
+        # Relative score (Normalized) reranker fusion type
+        else:
+            rerank = HybridFusion.RELATIVE_SCORE
+
+        # Perform hybrid search using Weaviate's native hybrid search
+        results = collection.query.hybrid(
```
Comment (franciscojavierarceo) [inline llama_stack/providers/remote/vector_io/weaviate/weaviate.py]: Again, please add try and except block here.

File: llama_stack/providers/remote/vector_io/weaviate/weaviate.py [hunk 6]
```diff
@@ -247,7 +358,7 @@ async def register_vector_db(
                 ],
             )
 
-        self.cache[sanitized_collection_name] = VectorDBWithIndex(
+        self.cache[vector_db.identifier] = VectorDBWithIndex(
```
Comment (franciscojavierarceo) [inline llama_stack/providers/remote/vector_io/weaviate/weaviate.py:372]: Looks like this was a bug?

File: tests/integration/vector_io/test_vector_io.py [hunk 7]
```diff
@@ -125,8 +125,9 @@ def test_insert_chunks(client_with_empty_registry, embedding_model_id, embedding
 def test_insert_chunks_with_precomputed_embeddings(client_with_empty_registry, embedding_model_id, embedding_dimension):
     vector_io_provider_params_dict = {
         "inline::milvus": {"score_threshold": -1.0},
-        "remote::qdrant": {"score_threshold": -1.0},
```
Comment (franciscojavierarceo) [inline tests/integration/vector_io/test_vector_io.py:141]: Strange this change came through

File: tests/unit/providers/vector_io/remote/test_weaviate.py [hunk 8]
```diff
@@ -0,0 +1,269 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (franciscojavierarceo) [inline tests/unit/providers/vector_io/remote/test_weaviate.py]: I don't think we need these unit tests. These should be covered by our integration tests that are generic enough but comprehensive enough to be easily extended by new providers and to cover our use cases. If not, we should enhance the integration tests to cover more edge cases.

Let me know if I'm missing something though.

File: tests/unit/providers/vector_io/conftest.py [hunk 9]
```diff
@@ -448,6 +450,71 @@ async def mock_query_chunks(vector_db_id, query, params=None):
...
+    return db_path
+
+
+@pytest.fixture
+async def weaviate_vec_index(weaviate_vec_db_path, embedding_dimension):
+    import pytest_socket
+    import weaviate
+
+    pytest_socket.enable_socket()
+    client = weaviate.connect_to_embedded(
```
Comment (r3v5) [inline tests/unit/providers/vector_io/conftest.py:465]: Hey Christian. Nice work, well done! Just one suggestion, may be it's a good idea to mock a connection to Weaviate DB to not require users to set up local Weaviate to run tests?

File: tests/unit/providers/vector_io/conftest.py [hunk 10]
```diff
@@ -448,6 +450,71 @@ async def mock_query_chunks(vector_db_id, query, params=None):
...
+
+
+@pytest.fixture
+async def weaviate_vec_adapter(weaviate_vec_db_path, mock_inference_api, embedding_dimension):
+    import pytest_socket
+    import weaviate
+
+    pytest_socket.enable_socket()
+
+    client = weaviate.connect_to_embedded(
```
Comment (r3v5) [inline tests/unit/providers/vector_io/conftest.py:485]: same here

**Comment by 2000krysztof** [comment]
My suggestions were addressed so lgtm. Very nice work.

File: llama_stack/providers/remote/vector_io/weaviate/weaviate.py (hunk 3 above)
Comment (ChristianZaccaria) [reply llama_stack/providers/remote/vector_io/weaviate/weaviate.py:51]: This is to use Weaviate client v4. The v3 client is deprecated:
```
@deprecated(
    """
Python client v3 `weaviate.Client(...)` has been removed.

Upgrade your code to use Python client v4 `weaviate.WeaviateClient` connections and methods.
```

File: llama_stack/providers/remote/vector_io/weaviate/weaviate.py (hunk 4 above)
Comment (ChristianZaccaria) [reply llama_stack/providers/remote/vector_io/weaviate/weaviate.py]: Added, good call, thank you!

File: llama_stack/providers/remote/vector_io/weaviate/weaviate.py (hunk 5 above)
Comment (ChristianZaccaria) [reply llama_stack/providers/remote/vector_io/weaviate/weaviate.py]: Good call, thank you!

File: tests/integration/vector_io/test_vector_io.py (hunk 7 above)
Comment (ChristianZaccaria) [reply tests/integration/vector_io/test_vector_io.py:141]: This change here added `remote::weaviate`, and manually re-ordered in alphabetical order.

File: tests/unit/providers/vector_io/remote/test_weaviate.py (hunk 8 above)
Comment (ChristianZaccaria) [reply tests/unit/providers/vector_io/remote/test_weaviate.py]: Yeah, this makes sense to me. All the unit tests I had are well covered by the integration tests. I removed them now, thanks!

File: tests/unit/providers/vector_io/conftest.py (hunk 9 above)
Comment (ChristianZaccaria) [reply tests/unit/providers/vector_io/conftest.py:465]: Hey Ian, I thought of that too, thanks for reviewing and the suggestion! :) .  I think using `connect_to_embedded` has several great benefits as opposed to using a mock. To explain, Weaviate has an **embedded mode** that works a bit similar to how faiss works. It doesn't start a docker container, but rather it runs 'in-process'. The user won't be required to setup a local Weaviate as this in-process server is created and connected to automatically. Moreover, for the integration test this provides more accurate and reliable results while using a lightweight 'in-process' db server. - Note: Not all vector_io providers have this capability, Weaviate does.

File: tests/unit/providers/vector_io/conftest.py (hunk 10 above)
Comment (ChristianZaccaria) [reply tests/unit/providers/vector_io/conftest.py:485]: Please see comment above, thanks!!

File: llama_stack/providers/remote/vector_io/weaviate/weaviate.py (hunk 6 above)
Comment (ChristianZaccaria) [reply llama_stack/providers/remote/vector_io/weaviate/weaviate.py:372]: Yes, I believe this was a bug. Since we just added Weaviate to the integration tests in this PR, this wasn't caught until now.

File: tests/unit/providers/vector_io/conftest.py (hunk 9 above)
Comment (r3v5) [reply tests/unit/providers/vector_io/conftest.py:465]: Great! Thanks for detailed explanation!

**Comment by r3v5** [comment]
lgtm

File: llama_stack/providers/remote/vector_io/weaviate/weaviate.py (hunk 3 above)
Comment (leseb) [reply llama_stack/providers/remote/vector_io/weaviate/weaviate.py:51]: Do we need a version pin or >= to get the v4 client?

Comment (ChristianZaccaria) [reply llama_stack/providers/remote/vector_io/weaviate/weaviate.py:51]: Correct, it's pinned in the `pyproject.toml` file under dependencies as:
```
"weaviate-client>=4.16.5"
```

Changes have been propagated to the `uv.lock` file.

File: pyproject.toml [hunk 11]
```diff
@@ -43,12 +43,13 @@ dependencies = [
...
-    "uvicorn>=0.34.0",                        # server
-    "opentelemetry-sdk>=1.30.0",              # server
+    "python-multipart>=0.0.20", # For fastapi Form
+    "uvicorn>=0.34.0", # server
+    "opentelemetry-sdk>=1.30.0", # server
     "opentelemetry-exporter-otlp-proto-http>=1.30.0", # server
-    "aiosqlite>=0.21.0",                      # server - for metadata store
-    "asyncpg",                                # for metadata store
+    "aiosqlite>=0.21.0", # server - for metadata store
+    "asyncpg", # for metadata store
```
Comment (leseb) [inline pyproject.toml]: this should be in the provider registry instead. llama_stack/providers/registry/vector_io.py

**Review by leseb — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 11 above)
Comment (leseb) [reply pyproject.toml]: wrong line, but the weaviate addition shouldn't be there

Comment (ChristianZaccaria) [reply pyproject.toml]: Good point, I've moved this now into `llama_stack/providers/registry/vector_io.py`. Thanks!

File: pyproject.toml [hunk 12]
```diff
@@ -25,8 +25,8 @@ classifiers = [
 ]
 dependencies = [
     "aiohttp",
-    "fastapi>=0.115.0,<1.0",                  # server
-    "fire",                                   # for MCP in LLS client
+    "fastapi>=0.115.0,<1.0", # server
```
Comment (leseb) [inline pyproject.toml]: please revert the changes here and in uv.lock

**Comment by ChristianZaccaria** [comment]
Weird, some unrelated tests failed, seem like a flake. Will push again to test.

Edit: The tests passed now.

File: pyproject.toml (hunk 12 above)
Comment (leseb) [reply pyproject.toml]: still there :)

Comment (ChristianZaccaria) [reply pyproject.toml]: Apologies, I thought I reverted both 🤦‍♂️ . Fixed now

File: tests/unit/providers/vector_io/remote/test_milvus.py [hunk 13]
```diff
@@ -324,3 +324,6 @@ async def test_query_hybrid_search_default_rrf(
     call_args = mock_milvus_client.hybrid_search.call_args
     ranker = call_args[1]["ranker"]
     assert ranker is not None
+
+
+# TODO: Write tests for the MilvusVectorIOAdapter class.
```
Comment (leseb) [inline tests/unit/providers/vector_io/remote/test_milvus.py]: please create a followup issue for this

File: llama_stack/providers/remote/vector_io/weaviate/weaviate.py [hunk 14]
```diff
@@ -88,14 +89,30 @@ async def delete_chunks(self, chunks_for_deletion: list[ChunkForDeletion]) -> No
...
+        Performs vector search using Weaviate's built-in vector search.
+        Args:
+            embedding: The query embedding vector
+            k: Limit of number of results to return
+            score_threshold: Minimum similarity score threshold
+        Returns:
+            QueryChunksResponse with chunks and scores
+        """
+        logger.info(
+            f"WEAVIATE VECTOR SEARCH CALLED: embedding_shape={embedding.shape}, k={k}, threshold={score_threshold}"
```
Comment (leseb) [inline llama_stack/providers/remote/vector_io/weaviate/weaviate.py:102]: is this a debug log?

File: llama_stack/providers/remote/vector_io/weaviate/weaviate.py [hunk 15]
```diff
@@ -105,16 +122,20 @@ async def query_vector(self, embedding: NDArray, k: int, score_threshold: float)
...
+                continue
+            # Convert cosine distance ∈ [0,2] -> normalized cosine similarity ∈ [0,1]
+            score = 1.0 - (float(doc.metadata.distance) / 2.0)
             if score < score_threshold:
                 continue
 
             chunks.append(chunk)
             scores.append(score)
 
+        logger.info(f"WEAVIATE VECTOR SEARCH RESULTS: Found {len(chunks)} chunks with scores {scores}")
```
Comment (leseb) [inline llama_stack/providers/remote/vector_io/weaviate/weaviate.py]: ditto

File: llama_stack/providers/remote/vector_io/weaviate/weaviate.py [hunk 16]
```diff
@@ -34,7 +35,7 @@
 
 from .config import WeaviateVectorIOConfig
 
-log = get_logger(name=__name__, category="vector_io::weaviate")
+logger = get_logger(name=__name__, category="vector_io::weaviate")
```
Comment (leseb) [inline llama_stack/providers/remote/vector_io/weaviate/weaviate.py]: this is change is unnecessary.

File: tests/unit/providers/vector_io/remote/test_milvus.py (hunk 13 above)
Comment (ChristianZaccaria) [reply tests/unit/providers/vector_io/remote/test_milvus.py]: I've removed the comment as all test cases for `MilvusVectorIOAdapter` class are covered in the integration tests. 

Based on @franciscojavierarceo's [comment](https://github.com/llamastack/llama-stack/pull/3264#discussion_r2312779756), maybe I can instead create an issue to add any edge cases from the `test_milvus.py` file to the integration tests and delete this file, if that sounds good.

I.e., the test contains some fallback mechanisms for keyword search that are not included in the integration tests.

File: llama_stack/providers/remote/vector_io/weaviate/weaviate.py (hunk 14 above)
Comment (ChristianZaccaria) [reply llama_stack/providers/remote/vector_io/weaviate/weaviate.py:102]: I've set them as debug logs, thanks!

File: llama_stack/providers/remote/vector_io/weaviate/weaviate.py (hunk 16 above)
Comment (ChristianZaccaria) [reply llama_stack/providers/remote/vector_io/weaviate/weaviate.py]: Reverted, thanks!

File: tests/unit/providers/vector_io/remote/test_milvus.py [hunk 17]
```diff
@@ -23,13 +23,13 @@
 with patch.dict("sys.modules", {"pymilvus": pymilvus_mock}):
     from llama_stack.providers.remote.vector_io.milvus.milvus import MilvusIndex
 
-# This test is a unit test for the MilvusVectorIOAdapter class. This should only contain
+# This test is a unit test for the MilvusIndex class. This should only contain
```
Comment (leseb) [inline tests/unit/providers/vector_io/remote/test_milvus.py]: How does this relate to the kw search for Weaviate?

File: tests/unit/providers/vector_io/remote/test_milvus.py [hunk 18]
```diff
@@ -23,13 +23,13 @@
...
 # tests which are specific to this class. More general (API-level) tests should be placed in
 # tests/integration/vector_io/
 #
 # How to run this test:
 #
-# pytest tests/unit/providers/vector_io/test_milvus.py \
+# pytest tests/unit/providers/vector_io/remote/test_milvus.py \
```
Comment (leseb) [inline tests/unit/providers/vector_io/remote/test_milvus.py]: ditto

File: tests/unit/providers/vector_io/remote/test_milvus.py (hunk 17 above)
Comment (ChristianZaccaria) [reply tests/unit/providers/vector_io/remote/test_milvus.py]: It doesn't at all, I just noticed the comment was inaccurate. I can fix that on a follow-up PR if preferred

File: llama_stack/providers/remote/vector_io/weaviate/weaviate.py [hunk 19]
```diff
@@ -136,7 +157,50 @@ async def query_keyword(
         k: int,
         score_threshold: float,
     ) -> QueryChunksResponse:
-        raise NotImplementedError("Keyword search is not supported in Weaviate")
+        """
+        Performs BM25-based keyword search using Weaviate's built-in full-text search.
```
Comment (franciscojavierarceo) [inline llama_stack/providers/remote/vector_io/weaviate/weaviate.py:161]: maybe i'm missing something but it looks like this doc string corresponds to the delete method, is that right?

Comment (ChristianZaccaria) [reply llama_stack/providers/remote/vector_io/weaviate/weaviate.py:161]: I don't think so, unless I'm misunderstanding. The doc string `Performs BM25-based keyword search using Weaviate's built-in full-text search` corresponds to the keyword search method. 

The `delete` method has its own doc string.

Comment (franciscojavierarceo) [reply llama_stack/providers/remote/vector_io/weaviate/weaviate.py:161]: oh weird the rendering was hiding the `query_keyword` definition. thanks!

Comment (franciscojavierarceo) [reply llama_stack/providers/remote/vector_io/weaviate/weaviate.py:161]: Image

just had to expand it to confirm. weird.

Comment (ChristianZaccaria) [reply llama_stack/providers/remote/vector_io/weaviate/weaviate.py:161]: Ohh you're right, that is weird. Looks like some GitHub bug 🤔

**Comment by ChristianZaccaria** [comment]
Rebased, I think we are ready to merge.
cc: @leseb

**Review by franciscojavierarceo — APPROVED** [review]
lgtm, thanks for this @ChristianZaccaria!
